### DISCUSSION TITLE
DevOpsTools also getting deployed

### DISCUSSION THREAD

**Discussion by Berns-s** [Cyfrin/foundry-full-course-cu#590] [Q&A]
When I use the get_most_recent_deployment of DevOpsTools, the DevOpsTools contract also get's deployed.  I use it in my deploy scripts. This is in run-latest,json ` "transactions": [
    {
      "hash": "0x5c127cfedd12b80223a54964d03eea96973cb8da24c4cbe77a951151e31465e7",
      "transactionType": "CREATE",
      "contractName": "DevOpsTools",
      "contractAddress": "0x8c47A626288E591Ddeb5600f786123F1D59b6003",
      "function": null,
      "arguments": null,
      "transaction": {
        "type": "0x02",
        "from": "0xadc2faccabcece89f3d4b15d361fd3e917c8e11f",
        "gas": "0xec9e3",
        "data": `

This is my run function in my script 
`function run() external {
   address mostRecentlyDeployed_Linker = DevOpsTools.get_most_recent_deployment("ContractLinker", 11155111);

             HelperConfig helperConfig = new HelperConfig();
            IERC20 usdcAddress = helperConfig.activeNetworkConfig();

              contractLinker = ContractLinker(mostRecentlyDeployed_Linker);
            vm.startBroadcast();
              castleBattle = new CastleBattle();

  castleBattle.initializeContractLinkerState(mostRecentlyDeployed_Linker);
        contractLinker.SetBattleCastleAddress(Strings.toHexString(uint256(uint160(address(castleBattle))), 20));
 
          vm.stopBroadcast();
    }`

**Comment by theirrationalone** [comment]
`ContractLinker` isn't deployed on local chain i.e. anvil, And DevOpsTools can only access locally deployed contracts which are stored in broadcast folder in json format by  reading contract's contractName inside json.

But You're using `11155111` Sepolia chainid(it will only try to retrieve data from broadcast folder/11155111/latest.json) and It is not possible for DevOpsTools to retrieve data from etherscan!

Hope it helps and Please Feel free to Correct me if somehow I went wrong.

Thank You!

**Reply by shikhar229169** [reply]
Idk why but actually DevOpsTools contract is also deployed, I also wonder that why this happens, there is no broadcast added for that, even then it also gets deployed. And also when Patrick deployed the contract and in his case I saw when all contracts got deployed, when he opened one of the contract on sepolia etherscan it was of DevOpsTool.

**Reply by theirrationalone** [reply]
This is horrible 😱😱😱... , @shikhar229169 are you sure about that devopstool contract also got deployed on real testnet ??? 

If yes then it would be terrifying because it's just for development purposes only, A tool only. 

Wonder wonder wonder. 

Hey @PatrickAlphaC , please Help us here....

**Reply by shikhar229169** [reply]
There you go.
https://youtu.be/wUjYK5gwNZs?t=21815

**Reply by theirrationalone** [reply]
Hey @shikhar229169 [look at this](https://github.com/Cyfrin/foundry-devops/blob/main/src/DevOpsTools.sol#L88) It's our own Library✅(Not Contract ❌) something === that PriceFeeds Library, nothing to manipulate/intreract in `testnets`|`fork`<==>`mainnet`.

Now explore that `get_most_recent_deployment`  function... We were worrying for nothing😅😅.... You Can also check its .sh below that Library(Edited)😅.

Actually We're deploying it for Local support only. Have you noticed, When we deploy a contract on a real `testnet` with `--broadcast` command, A `.json` file of that contract is also created and get stored into broadcast folder. So It's just a support tool for if we want to interact with our contracts directly from terminal or scripts, Then **Foundry's DevOpsTools Library** helps us to make interactions faster & reliable. It's like CI/CD (But in local-to-local) development.

**Reply by shikhar229169** [reply]
But the thing is why its getting deployed to testnet, only those txns are performed for real for which we have added vm.startBroadcast.
Like HelperConfig it was just a simulation but not an actual deployment.
Then why DevOpsTool is getting deployed.
See, the below addresses on sepolia:
0xc86767fB874d0bB3a9ECe0A9b9a7b1ef0cCaCa62
0xf316655d0ebf88f3d641aBC034E622aB75f8131D
0x52a9B97Eb402B7578a75187c64b1c2e9a7cF52b6
0x589cC43241d63fc4603dB335C5A1D9c8BdD021e5

**Reply by theirrationalone** [reply]
Haven't you noticed whilst working with contracts... ?? 


I'm taking @alfheimrShiven 's contract's example... 

[Check it out](https://sepolia.etherscan.io/address/0x5C7Ba1b0C0460c511203C9f2b3D37D1A5134d07d#code) 

All the contracts that are into `src` folder also get deployed as interdependent contracts. (But for onChain interactions) no matter where they were allocated? Whether into lib(dependencies) Or into main `src` (our folder) 


So the fact is, look on shiven's contract, He has imported other contracts into his contract... So here he's importing other contracts that means his contract can interact with other deployed contract on Real/Fork test/main network. 

And we also shiven too doesn't import `devOps` into main contract ever. (I'm sure, Now You're smart enough therefore You know what the spaghetti is...) 

There's no work for `devOps` into our main/src contracts.

And i noticed these overpowered features of foundry... 

1. Src: whatever is in src folder will get deployed. 

2. testRevert: if we try to test something with this function naming convention `function testRevertSomething` then foundry automatically expects revert without the need of `vm.expectRevert`.

3 . DeployFoo: foundry treats all files with this naming convention as special scripts files. Foundry isolates & separates files from testing files. Therefore we can't test files whose names start with Deploy. 

I'm confused 😵 in 3rd one. (Would you like to correct me??) 

Please correct me i went anywhere wrong...

**Reply by shikhar229169** [reply]
About your 3rd point:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/111875681/fb345f38-ac2d-496f-9121-084c3439ccb9)
You can see the test file name starts with Deploy

**Reply by theirrationalone** [reply]
Thanks to You(**Shikhar**)..... For clearing my Confusion, Now i'm aware of that behaviour too. ~ Meow 😼

**Comment by Berns-s** [comment]
I need more clarification, this is the output after running my deploy script, Only 1 contract should get deployed as in my deploy script above. The top contract is DevOpsTools and it states the gas paid. 
`Waiting for receipts.
⠂ [00:00:14] [##############################################################################################################################################################################################################] 5/5 receipts (0.0s)
##### sepolia
✅ Hash: 0x2a0bc091a58ab50afebd56f9a2f22d344c709b8fc3b250ab9fc57457295f3ddc
Contract Address: 0x66d02DD2D60478e3DB58b898972E80AB2215009c
Block: 4061305
Paid: 0.00238517858005631 ETH (745733 gas * 3.19843507 gwei)


##### sepolia
✅ Hash: 0x5fea391410d7098fdc1dff143fe69a981c9c3376313bbcad30dc03b791d7935f
Contract Address: 0x0F9B3b84A7adCd99C3b0152a62d06584d8f57bBB
Block: 4061305
Paid: 0.01254619100088721 ETH (3922603 gas * 3.19843507 gwei)`

**Reply by theirrationalone** [reply]
It matters in local development only.... 

In the mainnet we don't need anything from devops tool. So count that gas only once⛽

**Reply by Berns-s** [reply]
I do need to use it when I deploy to testnet/mainnet, 

I use to to update other previously deployed contracts with the newly deployed contracts address.

**Reply by theirrationalone** [reply]
But interacting through foundry.... Again that CI/CD thing i discussed with @shikhar229169 on above replies, 

Just imagine like this..... 

# What we're not doing using DevOps tool:

We need to let our Raffle/Lottery interact with chainlink vrf & keepers(Automation) using subscriptions, On Test/Main Network and we depict this behaviour by the help of "Mocks" So called Mocking(as You know what Mocking is!) For our local development. 

# What we're doing using Devops tool:

Like so, Thanks to **Cyfrin Team** for making local development way faster ⚡⚡, As a tool they made Foundry DevOps 🛠️ . 

Just to keep track of recent deployments of any network, we have that devOps 🛠️ as a facility to get recently deployed contracts. But if we won't have that tool then we would have to hardcode that recently deployed contract's name, address etc, each time to interact with any deployed contract and we won't be able to automate that interaction. 

And yes yes yes count Gas ⛽ each time when we deploy contracts with DevOps to `testnet/mainnet`.

Deploying a contract is a one time process because it's immutable And transparent. 

### Note: If you wish to not deploy DevOpsTools then, You can! 

I meant this with **ONCE**.... There's no role of DevOps on real testnet/mainnet.

**Comment by PatrickAlphaC** [comment]
Yep, this is an issue with foundry scripts at the moment :(

**Reply by rizoom** [reply]
Issue is still open but FWIW, declaring the library method as internal is a workaround. See comment: https://github.com/foundry-rs/foundry/issues/3295#issuecomment-1427635968

**Comment by theirrationalone** [comment]
> Yep, this is an issue with foundry scripts at the moment :(

SIR!, Is it something to worry about??

**Reply by Berns-s** [reply]
Probably not, you will just spend the gas for deploying the DevOpsTool with your contract.

**Reply by theirrationalone** [reply]
Although one time cost matters also 😥, 2-5 $ to deploy a useless protocol with upto 200+(appx) lines of code.
